### PR TITLE
[AMD-AIE][Objectfifo] Add e2e CI test for small shaped bf16 and i8 Matmul + lit test in lower-to-aie pass

### PR DESCRIPTION
**Pull request by Abhishek-Varma**
This PR contains 2 additions :-

-- Since because of [this commit in mlir-aie](https://github.com/erwei-xilinx/mlir-aie/commit/0504f7ad3662ce9a2044ea38f3a7f3df9b6708fc) we can now create `aiex.npu.dma_memcpy_nd` without having to linearize the arguments, I've added a lit test in `lower-to-aie` pass.
-- I've also added an e2e CI test for small shaped bf16 and i8 Matmul on Objectfifo.

Signed-off-by: Abhishek Varma <abhvarma@amd.com>

### CODE REVIEW

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELowerToAIE.cpp
Comment (nirvedhmeshram) [inline compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELowerToAIE.cpp:1018]: I will do a closer review later in the day but justa quick drive by comment, I think you want all the `2` used in divisions to be the variable `div`

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELowerToAIE.cpp
Comment (yzhang93) [inline compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELowerToAIE.cpp:1007]: I think it should be `oneDoffset = 0` and
`oneDoffset += offsets[j] * strides[j] for j=0 to (dmaUser.getMixedOffsets().size() - 1)` ?

Here the innermost stride is always 1?

**Review by yzhang93 — CHANGES_REQUESTED** [review]

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELowerToAIE.cpp
Comment (nirvedhmeshram) [inline compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELowerToAIE.cpp:992]: The memref rank changing  should not determine what we do to the offsets (same with sizes/strides)  is independent of the rank. I believe we need to divide any stride 1 offsets by `div` and can leave everything as is?

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/test/lower_to_aie.mlir
Comment (nirvedhmeshram) [inline compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/test/lower_to_aie.mlir:507]: This test is for offset all zeros could you makea another  tests that has non zero offsets?

File: build_tools/ci/run_matmul_test.sh
Comment (Abhishek-Varma) [reply build_tools/ci/run_matmul_test.sh:798]: Hi @yzhang93,

Yes, you're indeed correct.

I've logged [128x128x256 findings here](https://github.com/nod-ai/iree-amd-aie/issues/431#issuecomment-2197181642).

The intention was then to first add support in order to make basic/small shape bf16 Matmul work on objectfifo.

With @jtuyls' support for varied shapes on i32 and after this PR lands, we can take a look at supporting other shapes for bf16 incrementally.

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELowerToAIE.cpp
Comment (Abhishek-Varma) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELowerToAIE.cpp:1037]: I tried applying that logic but it didn't work.

This is the first row :-
```
64x64xf32=[384 384 384 384 384 384 384 384 384 384 384 384 384 384 384 384 384 384 384 384 384 384 384 384 384 384 384 384 384 384 384 384 256 256 256 256 256 256 256 256 256 256 256 256 256 256 256 256 256 256 256 256 256 256 256 256 256 256 256 256 256 256 256 256]
```

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELowerToAIE.cpp
Comment (jtuyls) [inline compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELowerToAIE.cpp:1024]: Why specifically check for 16 and 8? Wouldn't this work for 4 bits as well?

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELowerToAIE.cpp
Comment (jtuyls) [inline compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELowerToAIE.cpp:1027]: I wouldn't use `div` as a name as it's a std function.

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/test/lower_to_aie.mlir
Comment (nirvedhmeshram) [inline compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/test/lower_to_aie.mlir:671]: can you explain how you are reaching 34?, in the base example, you have 
` [%c0, %c1, %c2] [%c2, %c32, %c16] [%c16, %c32, %c1] `
this to me means linear offset is, 1*32+2*16 = 48, so now since you are stepping 2 elements, you get 24?

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELowerToAIE.cpp
Comment (nirvedhmeshram) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELowerToAIE.cpp:1037]: to me it looks like you are doing nothing to the offsets in the linearization, so I wonder if leaving them unmodified, would work?

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELowerToAIE.cpp
Comment (nirvedhmeshram) [inline compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELowerToAIE.cpp:1055]: I dont follow why we need this, I thought the following for loop will take care of dividing the size of the dim with unit sride?

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELowerToAIE.cpp
Comment (nirvedhmeshram) [inline compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELowerToAIE.cpp:1046]: actually I dont follow why we are restricted to make 4 dims always, we can just keep the same dims as the op already has?

File: build_tools/ci/run_matmul_test.sh
Comment (yzhang93) [reply build_tools/ci/run_matmul_test.sh:798]: The log is quite old. Does it still have the same error? I'm asking because I'm not sure if the casting you implemented in this PR work in general or just work because the starting offsets are all 0.

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/test/lower_to_aie.mlir
Comment (yzhang93) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/test/lower_to_aie.mlir:656]: This is also my concern.

File: build_tools/ci/run_matmul_test.sh
Comment (Abhishek-Varma) [reply build_tools/ci/run_matmul_test.sh:798]: Oh yes yes - it's still the same in `AIEAssignBufferAddresses` pass.

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELowerToAIE.cpp
Comment (Abhishek-Varma) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELowerToAIE.cpp:1024]: Yes, it should.

But since `AIRRtToNpu` kept this guarding for `8` and `16` bitwidth only, I didn't change - but we should be able to modify this gradually.

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELowerToAIE.cpp
Comment (Abhishek-Varma) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELowerToAIE.cpp:1046]: This is majorly because of restrictions in the lower stack. Here is [aiex.npu.dma_memcpy_nd's verify()](https://github.com/Xilinx/mlir-aie/blob/main/lib/Dialect/AIEX/IR/AIEXDialect.cpp#L187-L193) - it makes certain assumptions on size/stride and ideally we would want all the metadata to be of same shape.

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/test/lower_to_aie.mlir
Comment (Abhishek-Varma) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/test/lower_to_aie.mlir]: So, I tried the offset suggestion (as per https://github.com/nod-ai/iree-amd-aie/pull/573#discussion_r1684537348) and it worked.

The point you made for the linearized offset to be 17 makes sense to me. Good catch! :)
I've therefore added => `original linearized offset / upcastDivisor` as the linearized offset modification and I've added that as part of the current revision's push. It worked too.

But both the above point "worked" because we are dealing with a zero offset case. We cannot verify the same for a non-zero case just yet because of https://github.com/nod-ai/iree-amd-aie/pull/573#discussion_r1684159453.

I believe we should revisit/modify the logic for metadata later when we're able to successfully compile a non-zero offset case for bf16 as it'd help us in verifying the implementation (CC: @yzhang93 ).

And currently since we have our first bf16 Matmul case working (I again tried locally with the set of changes that went in recently and affected the input IR), it'd be nice to unblock getting this in, in my opinion.

But if we have a metadata algorithm that we know will work irrespective and there's some other way that we can verify  the same - I'd be happy to know and include the same in this PR.

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELowerToAIE.cpp
Comment (jtuyls) [inline compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELowerToAIE.cpp:976]:
```suggestion
/// Utility to convert function arguments with memref element types' bitwidth
/// either 4, 8 or 16 to i32. This basically involves linearizing an N-d memref to
/// 1-d i32 memref and adjusting the offset/size/stride metadata.
```

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELowerToAIE.cpp
Comment (jtuyls) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELowerToAIE.cpp]: The check is good for me like this, but maybe it's useful to add tests for 8 and 4 bits as well, so at least we're sure this logic works for different bitwidths? 8 bit will for sure come up soon and 4 bit might come up as well for BW reasons.

**Comment by yzhang93** [comment]
@Abhishek-Varma I just talked to @erwei-xilinx, and he mentioned bitwidth 8 / 16 is currently supported in mlir-aie and thus upcasting is not needed. Can you explore and see if we can directly use bf16 in the control codes?

**Comment by erwei-xilinx** [comment]
> @Abhishek-Varma I just talked to @erwei-xilinx, and he mentioned bitwidth 8 / 16 is currently supported in mlir-aie and thus upcasting is not needed. Can you explore and see if we can directly use bf16 in the control codes?

Yeah here's one example: https://github.com/Xilinx/mlir-aie/blob/eef800b3bb4eb335c17390121ff8bcb3445cfa64/test/npu-xrt/add_21_i8_using_dma_op_with_padding/aie.mlir#L72

**Comment by yzhang93** [comment]
> > @Abhishek-Varma I just talked to @erwei-xilinx, and he mentioned bitwidth 8 / 16 is currently supported in mlir-aie and thus upcasting is not needed. Can you explore and see if we can directly use bf16 in the control codes?
> 
> Yeah here's one example: https://github.com/Xilinx/mlir-aie/blob/eef800b3bb4eb335c17390121ff8bcb3445cfa64/test/npu-xrt/add_21_i8_using_dma_op_with_padding/aie.mlir#L72

Thanks! CC @makslevental as well, in case he has some insights about this.

**Comment by Abhishek-Varma** [comment]
> @Abhishek-Varma I just talked to @erwei-xilinx, and he mentioned bitwidth 8 / 16 is currently supported in mlir-aie and thus upcasting is not needed. Can you explore and see if we can directly use bf16 in the control codes?
> Yeah here's one example: https://github.com/Xilinx/mlir-aie/blob/eef800b3bb4eb335c17390121ff8bcb3445cfa64/test/npu-xrt/add_21_i8_using_dma_op_with_padding/aie.mlir#L72

Hi @yzhang93 @erwei-xilinx  - sure! That's great! Must be a recent support in `mlir-aie` then because earlier it wasn't allowing `aiex.npu.dma_cpy_nd` to be created if bitwidth wasn't `32`. 

I'll take a look at this and see if we can get through objectfifo's lowering. Thanks!

**Comment by Abhishek-Varma** [comment]
Hi @yzhang93 @erwei-xilinx - it worked! No issue in creating `aiex.npu.dma_memcpy_nd` ops anymore in `lower-to-aie`. Thanks!

**Review by jtuyls — APPROVED** [review]
LGTM

**Review by yzhang93 — APPROVED** [review]
Great, thanks!

**Pull request merged**
